### PR TITLE
adding the forceSync flag for handling conflicting topic configs

### LINKED ISSUE #114: Consider adding forceSync=true for topic updates
# Desired Behavior

Add `forceSync=true` for updating of topic metadata.

# Benefits

* See https://github.com/homeaway/stream-registry/blob/bacfe0a5d61fa4102bd0ff521f80d60ff90257e2/core/src/main/java/com/homeaway/streamplatform/streamregistry/db/dao/impl/KafkaManagerImpl.java#L128-L134
* By adding `forceSync=true` , when users get the error, they can use the `forceSync=true` to get the value from the underlying broker, so that they can then update.

### PR DESCRIPTION
**Pull request by krishnatg**
# stream-registry PR

Adding flag to provide an option to override user-provided topic config in case of already existing conflicting configs on the underlying brokers (https://github.com/ExpediaGroup/stream-registry/issues/114)

- Fixes #114 

### Added
* Boolean flag `forceSync` which can be set by a user while upserting a stream to override the provided topic configs in the stream and continue using the existing topic configs on the brokers.

# PR Checklist Forms

- [x] CHANGELOG.md updated
- [x] Reviewer assigned
- [x] PR assigned (presumably to submitter)
- [x] Labels added (enhancement, bug, documentation)

### CODE REVIEW

File: infra-provider-kafka/src/main/java/com/homeaway/streamplatform/streamregistry/provider/impl/KafkaInfraManager.java [hunk 1]
```diff
@@ -254,8 +254,12 @@ void updateTopics(ZkUtils zkUtils, List<String> topicsToUpdate, Map<String, Stri
             // to exactly match downstream config when the stream-registry has not "onboarded" existing topic
             // for the first time.
 
-            // TODO Alternatively we can add a forceSync=true flag, ignoring any user provided info, and only updating SR with the underlying settings
-            //      We should probably do forceSync=true anyway, as it provides a simple way to keep things in sync (#114)
+            // Added a forceSync=true flag, ignoring any user provided info, and only updating SR with the underlying settings (#114)
+            if (forceSync) {
+                // NOTHING TO DO!
+                log.info("topic config mismatch for {} ignored. Input config={}, actual (retained) config={}", topic, topicConfigMap, actualTopicConfig);
+                continue;
```
Comment (cricket007) [inline infra-provider-kafka/src/main/java/com/homeaway/streamplatform/streamregistry/provider/impl/KafkaInfraManager.java]: What exactly is being "forced" if all you do is `continue`?

Comment (krishnatg) [reply infra-provider-kafka/src/main/java/com/homeaway/streamplatform/streamregistry/provider/impl/KafkaInfraManager.java]: Well, the flag is "forcing" the stream-registry to sync the given topic configs with what exists already on the underlying brokers.
I guess I can say that by `continue`ing here, the code is "forcing" the flow to not go ahead to the following exception generation ..

Comment (cricket007) [reply infra-provider-kafka/src/main/java/com/homeaway/streamplatform/streamregistry/provider/impl/KafkaInfraManager.java]: I put alternative names in #114

Comment (cricket007) [reply infra-provider-kafka/src/main/java/com/homeaway/streamplatform/streamregistry/provider/impl/KafkaInfraManager.java]: > ignoring any user provided info, and only updating SR with the underlying settings

The way I interpret that is to take `actualTopicConfig` and overwrite `topicConfigMap` with it...

**Comment by neoword** [comment]
Some clarification on `forceSync=true`:
- This should SUCCEED, IF AND ONLY IF config is THE SAME on all underlying clusters
- This should FAIL, if any of the cluster (stream-bindings) differ, with an error having a call to action to make fix the corresponding clusters that are different.

**Comment by neoword** [comment]
Waiting for more info from team before I comment on this PR.

File: infra-provider-kafka/src/main/java/com/homeaway/streamplatform/streamregistry/provider/impl/KafkaInfraManager.java (hunk 1 above)
Comment (neoword) [reply infra-provider-kafka/src/main/java/com/homeaway/streamplatform/streamregistry/provider/impl/KafkaInfraManager.java]: Doesn't `continue` actually perform the UPDATE???
I agree with @cricket007 remarks on `actualTopicConfig` and `topicConfigMap`... where we should be overriding what the user provided, with what is _actually_ out there.

**Comment by neoword** [comment]
Also should `forceSync=true` fail if no underlying topic exists?

File: infra-provider-kafka/src/main/java/com/homeaway/streamplatform/streamregistry/provider/impl/KafkaInfraManager.java (hunk 1 above)
Comment (cricket007) [reply infra-provider-kafka/src/main/java/com/homeaway/streamplatform/streamregistry/provider/impl/KafkaInfraManager.java]: > Doesn't continue actually perform the UPDATE?

No, it breaks here and goes to the next iteration of the loop. `updateTopic` isn't called.

Comment (neoword) [reply infra-provider-kafka/src/main/java/com/homeaway/streamplatform/streamregistry/provider/impl/KafkaInfraManager.java]: Oh. I missed a curly-brace close.

line 270 is INSIDE not... outside the loop.
https://github.com/ExpediaGroup/stream-registry/pull/170/files/09090b33c68b43bf046a36ddd7b7ed2959b849e3#diff-b4dcb632004addc439896ec21e5b7dabR270

**Comment by arunvasudevan** [comment]
Here are my comments on the PR:
1. `forceSync` is definitely a mis-guiding name..either `ignoreTopicCreationIfConfigsMismatch` or `skipTopicCreationIfConfigsMismatch` or something along that lines.

2. I can see this PR as a followup to the PR (https://github.com/ExpediaGroup/stream-registry/pull/52).
If the provided configs are different from the underlying topic config we have 2 options:
    - Force the underlying topic with the provided config  -> We don't want to do this based on the above PR (#52) and the issue that led to it.
    - Ignore (or skip) topic creation -> This is safer than the above option but the user may not notice the log in the CI/CD and it could lead to unexpected behavior later in the application.

Why not create topic only at the vpc and not in the replicatedVPC? creating topics in the replicatedVPC is a replication tools task isn't it?

**Comment by krishnatg** [comment]
Since (https://github.com/ExpediaGroup/stream-registry/pull/178) is merged, this PR is not required.

**Pull request closed without merging**